yoyodyne/data/datamodules.py

```diff
@@ -0,0 +1,200 @@
+        max_target_length: int = defaults.MAX_TARGET_LENGTH,
+    ):
+        super().__init__()
+        self.parser = tsv.TsvParser(
+            source_col=source_col,
+            features_col=features_col,
+            target_col=target_col,
+            source_sep=source_sep,
+            features_sep=features_sep,
+            target_sep=target_sep,
+        )
+        self.train = train
+        self.val = val
+        self.predict = predict
+        self.test = test
+        # Computes index.
+        source_vocabulary: Set[str] = set()
+        features_vocabulary: Set[str] = set()
+        target_vocabulary: Set[str] = set()
+        for path in self.paths:
+            if self.parser.has_features:
+                if self.parser.has_target:
+                    for source, features, target in self.parser.samples(path):
+                        source_vocabulary.update(source)
+                        features_vocabulary.update(features)
+                        target_vocabulary.update(target)
+                else:
+                    for source, features in self.parser.samples(path):
+                        source_vocabulary.update(source)
+                        features_vocabulary.update(features)
+            elif self.parser.has_target:
+                for source, target in self.parser.samples(path):
+                    source_vocabulary.update(source)
+                    target_vocabulary.update(target)
+            else:
+                for source in self.parser.samples(path):
+                    source_vocabulary.update(source)
+            if self.parser.has_target and tied_vocabulary:
+                source_vocabulary.update(target_vocabulary)
+                target_vocabulary.update(source_vocabulary)
+        self.separate_features = separate_features
+        self.index = indexes.Index(
+            source_vocabulary=sorted(source_vocabulary),
+            features_vocabulary=sorted(features_vocabulary)
+            if features_vocabulary
+            else None,
+            target_vocabulary=sorted(target_vocabulary)
+            if target_vocabulary
+            else None,
+        )
+        # Stores batch size.
+        self.batch_size = batch_size
+        # Makes collator.
+        self.collator = collators.Collator(
+            pad_idx=self.index.pad_idx,
+            has_features=self.index.has_features,
+            has_target=self.index.has_target,
+            separate_features=separate_features,
+            features_offset=self.index.source_vocab_size
+            if self.index.has_features
+            else 0,
+            max_source_length=max_source_length,
+            max_target_length=max_target_length,
+        )
+
+    # Helpers.
+
+    @property
+    def paths(self) -> Iterator[str]:
+        if self.train is not None:
+            yield self.train
+        if self.val is not None:
+            yield self.val
+        if self.predict is not None:
+            yield self.predict
```

Comment:
It isn't supported now; you'll get a parsing error. Same with `--test`.

I think with LightningCLI it will be technically supported but no-op, which also seems fine. (We could make it "op" if you want at a later point.)

Reply:
Sounds good. If we decide we want them "op", then that sounds like a separate PR.